### ISSUE TITLE
[BUG] - Playwright WebKit browser leaks IP while using a proxy

### ISSUE THREAD

**Issue by aemla** [davidteather/TikTok-Api#878]
Playwright WebKit browser leaks IP while using a proxy. It makes proxies useless because TikTok can detect your public IP address via WebRTC and for this reason TikTok shows captcha even with a new proxy and a different verify_fp.

Possible fix:
Somehow disable WebRTC to avoid leaking or use Firefox with firefoxUserPrefs={"media.peerconnection.enabled": False}

How to reproduce:
Open webkit browser and go to: https://browserleaks.com/webrtc

**Comment by SuperSonicHub1** [comment]
It seems that the solution to stopping IP leaking is to use "Enable Legacy WebRTC API": https://privacysavvy.com/security/safe-browsing/disable-webrtc-chrome-firefox-safari-opera-edge/#h-disabling-webrtc-on-safari

By enabling the option, it decimates the API surface fingerprinting attacks rely on: https://webkit.org/blog/7763/a-closer-look-into-webrtc/ (See "Legacy WebRTC and Media Streams API.")

We don't have a nice little toolbar with a headless browser, though. We do, however, have command line flags.
After doing a bit of searching, I found the WebKit patch that added the Legacy API flag: https://bugs.webkit.org/attachment.cgi?id=304704&action=diff#a/Source/WebKit2/Shared/WebPreferencesDefinitions.h_sec1
I hope it's as simple as --preferences WebRTCLegacyAPIEnabled=true, or something like that.

Sadly, entering "webkit command line options" into my search engine hasn't done me much good.

**Comment by SuperSonicHub1** [comment]
Found it!
```
--enable-webrtc=gboolean Whether WebRTC content should be handled
```
Will be pushing a patch ASAP. For future reference here's all of the MiniBrowser command-line args and how to access them.
```
/home/kwilliams/.cache/ms-playwright/webkit-1683/pw_run.sh --help-all
Usage:
  MiniBrowser [OPTION…] [URL…]

Help Options:
  -h, --help                                                      Show help options
  --help-all                                                      Show all help options
  --help-gtk                                                      Show GTK+ Options
  --help-gst                                                      Show GStreamer Options
  --help-websettings                                              WebKitSettings properties

GTK+ Options
  --class=CLASS                                                   Program class as used by the window manager
  --name=NAME                                                     Program name as used by the window manager
  --gdk-debug=FLAGS                                               GDK debugging flags to set
  --gdk-no-debug=FLAGS                                            GDK debugging flags to unset
  --gtk-module=MODULES                                            Load additional GTK+ modules
  --g-fatal-warnings                                              Make all warnings fatal
  --gtk-debug=FLAGS                                               GTK+ debugging flags to set
  --gtk-no-debug=FLAGS                                            GTK+ debugging flags to unset

GStreamer Options
  --gst-version                                                   Print the GStreamer version
  --gst-fatal-warnings                                            Make all warnings fatal
  --gst-debug-help                                                Print available debug categories and exit
  --gst-debug-level=LEVEL                                         Default debug level from 1 (only error) to 9 (anything) or 0 for no output
  --gst-debug=LIST                                                Comma-separated list of category_name:level pairs to set specific levels for the individual categories. Example: GST_AUTOPLUG:5,GST_ELEMENT_*:3
  --gst-debug-no-color                                            Disable colored debugging output
  --gst-debug-color-mode                                          Changes coloring mode of the debug log. Possible modes: off, on, disable, auto, unix
  --gst-debug-disable                                             Disable debugging
  --gst-plugin-spew                                               Enable verbose plugin loading diagnostics
  --gst-plugin-path=PATHS                                         Colon-separated paths containing plugins
  --gst-plugin-load=PLUGINS                                       Comma-separated list of plugins to preload in addition to the list stored in environment variable GST_PLUGIN_PATH
  --gst-disable-segtrap                                           Disable trapping of segmentation faults during plugin loading
  --gst-disable-registry-update                                   Disable updating the registry
  --gst-disable-registry-fork                                     Disable spawning a helper process while scanning the registry

WebKitSettings writable properties for default WebKitWebView
  --enable-javascript=gboolean                                    Enable JavaScript.
  --auto-load-images=gboolean                                     Load images automatically.
  --load-icons-ignoring-image-load-setting=gboolean               Whether to load site icons ignoring image load setting.
  --enable-offline-web-application-cache=gboolean                 Whether to enable offline web application cache.
  --enable-html5-local-storage=gboolean                           Whether to enable HTML5 Local Storage support.
  --enable-html5-database=gboolean                                Whether to enable HTML5 database support.
  --enable-xss-auditor=gboolean                                   Whether to enable the XSS auditor.
  --enable-frame-flattening=gboolean                              Whether to enable frame flattening.
  --enable-plugins=gboolean                                       Enable embedded plugin objects.
  --enable-java=gboolean                                          Whether Java support should be enabled.
  --javascript-can-open-windows-automatically=gboolean            Whether JavaScript can open windows automatically.
  --enable-hyperlink-auditing=gboolean                            Whether <a ping> should be able to send pings.
  --default-font-family=gchararray                                The font family to use as the default for content that does not specify a font.
  --monospace-font-family=gchararray                              The font family used as the default for content using monospace font.
  --serif-font-family=gchararray                                  The font family used as the default for content using serif font.
  --sans-serif-font-family=gchararray                             The font family used as the default for content using sans-serif font.
  --cursive-font-family=gchararray                                The font family used as the default for content using cursive font.
  --fantasy-font-family=gchararray                                The font family used as the default for content using fantasy font.
  --pictograph-font-family=gchararray                             The font family used as the default for content using pictograph font.
  --default-charset=gchararray                                    The default text charset used when interpreting content with unspecified charset.
  --enable-private-browsing=gboolean                              Whether to enable private browsing
  --enable-developer-extras=gboolean                              Whether to enable developer extras
  --enable-resizable-text-areas=gboolean                          Whether to enable resizable text areas
  --enable-tabs-to-links=gboolean                                 Whether to enable tabs to links
  --enable-dns-prefetching=gboolean                               Whether to enable DNS prefetching
  --enable-caret-browsing=gboolean                                Whether to enable accessibility enhanced keyboard navigation
  --enable-fullscreen=gboolean                                    Whether to enable the Javascript Fullscreen API
  --print-backgrounds=gboolean                                    Whether background images should be drawn during printing
  --enable-webaudio=gboolean                                      Whether WebAudio content should be handled
  --enable-webgl=gboolean                                         Whether WebGL content should be rendered
  --allow-modal-dialogs=gboolean                                  Whether it is possible to create modal dialogs
  --zoom-text-only=gboolean                                       Whether zoom level of web view changes only the text size
  --javascript-can-access-clipboard=gboolean                      Whether JavaScript can access Clipboard
  --media-playback-requires-user-gesture=gboolean                 Whether media playback requires user gesture
  --media-playback-allows-inline=gboolean                         Whether media playback allows inline
  --draw-compositing-indicators=gboolean                          Whether to draw compositing borders and repaint counters
  --enable-site-specific-quirks=gboolean                          Enables the site-specific compatibility workarounds
  --enable-page-cache=gboolean                                    Whether the page cache should be used
  --user-agent=gchararray                                         The user agent string
  --enable-smooth-scrolling=gboolean                              Whether to enable smooth scrolling
  --enable-accelerated-2d-canvas=gboolean                         Whether to enable accelerated 2D canvas
  --enable-write-console-messages-to-stdout=gboolean              Whether to write console messages on stdout
  --enable-media-stream=gboolean                                  Whether MediaStream content should be handled
  --enable-mock-capture-devices=gboolean                          Whether we expose mock capture devices or not
  --enable-spatial-navigation=gboolean                            Whether to enable Spatial Navigation support.
  --enable-mediasource=gboolean                                   Whether MediaSource should be enabled.
  --enable-encrypted-media=gboolean                               Whether EncryptedMedia should be enabled.
  --enable-media-capabilities=gboolean                            Whether MediaCapabilities should be enabled.
  --allow-file-access-from-file-urls=gboolean                     Whether file access is allowed from file URLs.
  --allow-universal-access-from-file-urls=gboolean                Whether or not universal access is allowed from the context of file scheme URLs
  --allow-top-navigation-to-data-urls=gboolean                    Whether or not top frame navigation is allowed to data URLs
  --enable-back-forward-navigation-gestures=gboolean              Whether horizontal swipe gesture will trigger back-forward navigation
  --enable-javascript-markup=gboolean                             Enable JavaScript in document markup.
  --enable-media=gboolean                                         Whether media content should be handled
  --media-content-types-requiring-hardware-support=gchararray     List of media content types requiring hardware support.
  --enable-webrtc=gboolean                                        Whether WebRTC content should be handled

Application Options:
  --autoplay-policy                                               Autoplay policy. Valid options are: allow, allow-without-sound, and deny
  --bg-color                                                      Background color
  -e, --editor-mode                                               Run in editor mode
  -d, --dark-mode                                                 Run in dark mode
  -s, --session-file=FILE                                         Session file
  -g, --geometry=GEOMETRY                                         Unused. Kept for backwards-compatibility only
  -f, --full-screen                                               Set the window to full-screen mode
  -p, --private                                                   Run in private browsing mode
  --automation                                                    Run in automation mode
  -c, --cookies-file=FILE                                         Persistent cookie storage database file
  --cookies-policy=POLICY                                         Cookies accept policy (always, never, no-third-party). Default: no-third-party
  --proxy=PROXY                                                   Set proxy
  --ignore-host=HOSTS                                             Set proxy ignore hosts
  --ignore-tls-errors                                             Ignore TLS errors
  --content-filter=FILE                                           JSON with content filtering rules
  --enable-itp                                                    Enable Intelligent Tracking Prevention (ITP)
  --enable-sandbox                                                Enable web process sandbox support
  --exit-after-load                                               Quit the browser after the load finishes
  -t, --time-zone=TIMEZONE                                        Set time zone
  -v, --version                                                   Print the WebKitGTK version
  --inspector-pipe                                                Open pipe connection to the remote inspector
  --user-data-dir                                                 Default profile persistence folder location
  --headless                                                      Noop headless operation
  --no-startup-window                                             Do not open default page
  --display=DISPLAY                                               X display to use
```

**Comment by SuperSonicHub1** [comment]
Wanted to confirm that WebRTC is indeed disabled; it is.
![Screen%20Shot%202022-07-26%20at%2010 40 06](https://user-images.githubusercontent.com/25624882/181036248-8677adf8-8ec2-4a60-bbc1-baa278579368.png)

**Comment by SuperSonicHub1** [comment]
I don't completely understand why, but adding MiniBrowser seems to be unable to parse `--enable-webrtc=false` or `--enable-webrtc=0`. 
```
E       playwright._impl._api_types.Error: Browser closed.
E       ==================== Browser output: ====================
E       <launching> /home/kwilliams/.cache/ms-playwright/webkit-1683/pw_run.sh --inspector-pipe --headless --no-startup-window --enable-webrtc=false
E       <launched> pid=347460
E       [pid=347460][err] Cannot parse arguments: Unknown option --enable-webrtc=false
E       =========================== logs ===========================
E       <launching> /home/kwilliams/.cache/ms-playwright/webkit-1683/pw_run.sh --inspector-pipe --headless --no-startup-window --enable-webrtc=false
E       <launched> pid=347460
E       [pid=347460][err] Cannot parse arguments: Unknown option --enable-webrtc=false
E       ============================================================

env/lib/python3.10/site-packages/playwright/_impl/_connection.py:78: Error
```

`--enable-webrtc=false` works on the command line; otherwise I wouldn't have been able to post the screenshot in my previous comment.

`--enable-webrtc 0` also works from the command line, but Playwright itself, not the browser, isn't having it: `playwright._impl._api_types.Error: Arguments can not specify page to be opened`.

Any ideas?

**Comment by SuperSonicHub1** [comment]
https://github.com/microsoft/playwright/issues/15958
tl;dr: specifying `--headless` causes a different binary to be opened when running `pw_run.sh` which doesn't have `--enable-webrtc` as a command line flag. Y'know, Firefox is looking real nice right now.

**Comment by SuperSonicHub1** [comment]
I'm considering writing a patch which lets the user pick their browser engine. Then, you could pick Firefox and then supply `firefox_user_prefs`.

**Comment by davidteather** [comment]
Switched to chrome in V6 hopefully isn't leaking 🤞
